warehouse/malware/checks/setup_patterns/check.py

```diff
@@ -0,0 +1,120 @@
+from warehouse.malware.models import VerdictClassification, VerdictConfidence
+from warehouse.packaging.models import File
+
+
+class SetupPatternCheck(MalwareCheckBase):
+    _yara_rule_file = os.path.join(
+        os.path.dirname(os.path.abspath(__file__)), "setup_py_rules.yara"
+    )
+
+    version = 1
+    short_description = "A check for common malicious patterns in setup.py"
+    long_description = dedent(
+        """
+        This check uses YARA to search for common malicious patterns in the setup.py
+        files of uploaded release archives.
+        """
+    )
+    check_type = "event_hook"
+    hooked_object = "File"
+
+    def __init__(self, db):
+        super().__init__(db)
+        self._yara_rules = self._load_yara_rules()
+
+    def _load_yara_rules(self):
+        return yara.compile(filepath=self._yara_rule_file)
+
+    def _extract_setup_contents(self, archive_stream):
+        if zipfile.is_zipfile(archive_stream):
+            with zipfile.ZipFile(archive_stream) as zipobj:
+                for name in zipobj.namelist():
+                    parts = os.path.split(name)
+                    if len(parts) == 2 and parts[1] == "setup.py":
+                        return zipobj.read(name)
+                return None
+        else:
+            # NOTE: is_zipfile doesn't rewind the fileobj it's given.
+            archive_stream.seek(0)
+
+            # NOTE: We don't need to perform a sanity check on
+            # the (presumed) tarfile's compression here, since we're
+            # extracting from a stream that's already gone through
+            # upload validation.
+            # See _is_valid_dist_file in forklift/legacy.py.
+            try:
+                with tarfile.open(fileobj=archive_stream) as tarobj:
+                    member = tarobj.next()
+                    while member:
+                        parts = os.path.split(member.name)
+                        if len(parts) == 2 and parts[1] == "setup.py":
+                            return tarobj.extractfile(member).read()
+                        member = tarobj.next()
+                    return None
+            except tarfile.TarError:
+                return None
+
+    def scan(self, **kwargs):
+        file_id = kwargs.get("obj_id")
+        file_url = kwargs.get("file_url")
+        if file_id is None or file_url is None:
+            # TODO: Maybe raise here, since the absence of these
+            # arguments is a use/user error.
+            return
+
+        file = self.db.query(File).get(file_id)
+        if file.packagetype != "sdist":
+            # Per PEP 491: bdists do not contain setup.py.
+            # This check only scans dists that contain setup.py, so
+            # we have nothing to perform.
+            return
+
+        # TODO(ww): Handle connection errors, timeouts?
+        # The check harness will catch everything eventually,
+        # so these might not be worth handling explicitly.
```

Comment:
I agree with the approach outlined in this comment. Maybe remove the `file_request.ok` checking and let it raise an `AttributeError` a few lines down on `file_request.content` to that `run_check` retries a few times.

Reply:
Sounds good! I'm refactoring this into a general utility but I'll make it as noisy as possible (in terms of exceptions) to ensure that we propagate them outwards for retrying.